src/client/Client.cc

```diff
@@ -4150,6 +4150,7 @@ void Client::remove_cap(Cap *cap, bool queue_release)
   }
   size_t n = in.caps.erase(mds);
   ceph_assert(n == 1);
+  cap->remove_cap_item();
```

Comment:
`in.caps` is a `std::map<mds_rank_t, Cap>`. The `Cap` destructor will be called in `in.caps.erase(mds)` above. This `cap->remove_cap_item()` will have no effect.

I think the original bug is simply an issue of accounting. Not a real defect in the code.